### PR TITLE
the priority of H2D

### PR DESCRIPTION
**Pull request by zhaozheng09**
Question：https://github.com/tensorflow/tensorflow/issues/58616
Modify:
We distinguish FeaturesAsyncH2D and ComputeOpH2D, and raise the priority of ComputeOpH2D and lower the priority of FeaturesAsyncH2D.  In other words: 
Old: When the previous feature transfer has finished submitting, the next feature transfer can start.
New: Record FeaturesH2D launch size, and then when launch size reaches the threshold, we add a `BlockUntilHost` event to H2DStream to lower the priority of FeaturesAsync2D.
Result: ComputeOpH2D will execute immediately.
image

Advantage: ComputeOpH2D have higher priority, because FeaturesAsyncH2D will `BlockUntilHost` sometimes.
Disadvantage: 
1. Maybe use more gpu memory.
  a. Before: Each feature will create a new gpu tensor and each feature sync once in ComputeStream and H2DStream.
  b. After: All features batch create gpu tensors and All features sync once in ComputeStream and H2DStream.
  so we need more gpu memory.
3. The performance of FeaturesAsyncH2D will decrease.

**Due to busy work, if you find this feature useful, I will complete my PR later.**

### CODE REVIEW

**Comment by gbaned** [comment]
@zhaozheng09 Can you please resolve conflicts? Thank you!

**Comment by zhaozheng09** [comment]
> @zhaozheng09 Can you please resolve conflicts? Thank you!

Done,

**Comment by zhaozheng09** [comment]
@gbaned Hi, Do we need this functionality?

**Comment by zhaozheng09** [comment]
@gbaned Hi, Do we need this feature? and @zhangyujing seems to have been offline for a long time

**Comment by gbaned** [comment]
> what should I do now?

Hi @zhaozheng09 Sorry for the delay, nothing pending from your end now, it is awaiting for review. Thank you!

**Comment by gbaned** [comment]
Hi @zhangyujing Can you please review this PR ? Thank you!

File: tensorflow/core/common_runtime/gpu/gpu_util.cc [hunk 1]
```diff
@@ -371,9 +372,28 @@ void GPUUtil::CopyCPUTensorToGPU(const Tensor* cpu_tensor,
...
+      if (h2d_wait_self) {
+        static int64 transfer_size = 0;
+        transfer_size += total_bytes;
+        if (transfer_size > SyncSelfSizeOnce()) {
+          auto h2d_stream =
+              stream_executor::gpu::AsGpuStreamValue(recv_host_to_device_stream);
+          cudaStreamSynchronize(h2d_stream);
+          transfer_size = 0;
+        }
+    }
```
Comment (cantonios) [inline tensorflow/core/common_runtime/gpu/gpu_util.cc:384]: nit: indentation is off here - please run through clang-format.

**Review by cantonios — CHANGES_REQUESTED** [review]
I don't know much about this code.  @reedwm any thoughts here?

**Comment by reedwm** [comment]
It looks like this PR does two independent things:
  1. Synchronize the H2D stream every time the GPU receives a total of N bytes from the host (N is determined by the `SyncSelfSizeOnce()` function, which is not declared/defined by this PR)
  2. In `RecvOutputsFromRendezvousAsync`, batch allocate all tensors instead of allocating each separately. I'm not familiar with `RecvOutputsFromRendezvousAsync` and when it is called, so I'm not sure what impact this has.

For (1), I'm not sure how this raises the priority of the H2D stream, and also I am not sure why we want to raise its priority. If we do want to raise the priority, I think setting priorities with `cudaStreamCreateWithPriority` would be a better approach. For (2), I'm not sure what the purpose is. @zhaozheng09 can you clarify?

**Comment by zhaozheng09** [comment]
> It looks like this PR does two independent things:
> 
> 1. Synchronize the H2D stream every time the GPU receives a total of N bytes from the host (N is determined by the `SyncSelfSizeOnce()` function, which is not declared/defined by this PR)
> 2. In `RecvOutputsFromRendezvousAsync`, batch allocate all tensors instead of allocating each separately. I'm not familiar with `RecvOutputsFromRendezvousAsync` and when it is called, so I'm not sure what impact this has.
> 
> For (1), I'm not sure how this raises the priority of the H2D stream, and also I am not sure why we want to raise its priority. If we do want to raise the priority, I think setting priorities with `cudaStreamCreateWithPriority` would be a better approach. For (2), I'm not sure what the purpose is. @zhaozheng09 can you clarify?

For(1): ```cudaStreamCreateWithPriority ``` is only affects the kernels, it has no effect for H2D/D2H;
For(2): I should write a simple example ...

**Comment by gbaned** [comment]
Hi @reedwm Can you please take a look on the above [comment](https://github.com/tensorflow/tensorflow/pull/58617#issuecomment-1493206593) from @zhaozheng09. Thank you!

**Comment by gbaned** [comment]
Hi @reedwm Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @reedwm Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @reedwm Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @reedwm Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @reedwm Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @reedwm Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @reedwm Any update on this PR? Please. Thank you!

**Pull request closed without merging**